src/core/cpu_patches.cpp

```diff
@@ -0,0 +1,381 @@
+    // https://github.com/wine-mirror/wine/blob/a27b9551/dlls/kernelbase/thread.c#L719
+    static constexpr u32 TlsSlotsOffset = 0x1480;
+    static constexpr u32 TlsExpansionSlotsOffset = 0x1780;
+    static constexpr u32 TlsMinimumAvailable = 64;
+
+    const u32 teb_offset = slot < TlsMinimumAvailable ? TlsSlotsOffset : TlsExpansionSlotsOffset;
+    const u32 tls_index = slot < TlsMinimumAvailable ? slot : slot - TlsMinimumAvailable;
+
+    // Load the pointer to the table of TLS slots.
+    c.putSeg(Xbyak::util::gs);
+    c.mov(dst, Xbyak::util::ptr[reinterpret_cast<void*>(teb_offset)]);
+    // Load the pointer to our buffer.
+    c.mov(dst, Xbyak::util::qword[dst + tls_index * sizeof(LPVOID)]);
+#elif defined(__APPLE__)
+    // The following logic is based on the Darwin implementation of _os_tsd_get_direct, used by
+    // pthread_getspecific https://github.com/apple/darwin-xnu/blob/main/libsyscall/os/tsd.h#L89-L96
+    c.putSeg(Xbyak::util::gs);
+    c.mov(dst, Xbyak::util::qword[reinterpret_cast<void*>(slot * sizeof(void*))]);
+#else
+    // The following logic is based on the glibc implementation of pthread_getspecific
+    // https://github.com/bminor/glibc/blob/29807a27/nptl/pthread_getspecific.c#L23
+    static constexpr u32 PthreadKeySecondLevelSize = 32;
+    static constexpr u32 PthreadSpecificOffset = 0x510;
+
+    ASSERT(slot >= PthreadKeySecondLevelSize);
+
+    const u32 idx1st = slot / PthreadKeySecondLevelSize;
+    const u32 idx2nd = slot % PthreadKeySecondLevelSize;
+
+    // Load first level specific array.
+    c.putSeg(Xbyak::util::fs);
+    c.mov(dst, Xbyak::util::qword[reinterpret_cast<void*>(PthreadSpecificOffset + idx1st * 8)]);
+    // Load data member of pthread_key_data our slot specifies.
+    c.mov(dst, Xbyak::util::qword[dst + idx2nd * 16 + 8]);
+#endif
+
+    return true;
+}
+
+#ifdef __APPLE__
+
+static constexpr u32 MaxSavedRegisters = 3;
+static pthread_key_t register_save_slots[MaxSavedRegisters];
+static std::once_flag register_save_init_flag;
+
+static_assert(sizeof(void*) == sizeof(u64),
+              "Cannot fit a register inside a thread local storage slot.");
+
+static void InitializeRegisterSaveSlots() {
+    for (u32 i = 0; i < MaxSavedRegisters; i++) {
+        ASSERT_MSG(pthread_key_create(&register_save_slots[i], nullptr) == 0,
+                   "Unable to allocate thread-local register save slot {}", i);
+    }
+}
+
+static void SaveRegisters(Xbyak::CodeGenerator& c, const std::initializer_list<Xbyak::Reg> regs) {
+    ASSERT_MSG(regs.size() <= MaxSavedRegisters, "Not enough space to save {} registers.",
+               regs.size());
+
+    std::call_once(register_save_init_flag, &InitializeRegisterSaveSlots);
+
+    u32 index = 0;
+    for (const auto& reg : regs) {
+        const auto offset = reinterpret_cast<void*>(register_save_slots[index++] * sizeof(void*));
+
+        c.putSeg(Xbyak::util::gs);
+        c.mov(Xbyak::util::qword[offset], reg.cvt64());
```

Comment:
Why not push the registers to the stack instead? The GS segment could be used as guest TLS on unix platforms, avoiding the need for patches. I have done this in my linux fixes PR

Reply:
Originally this pushed registers to the stack, but you run into problems when guest code writes below the stack pointer without decrementing. I could try to guess a safe distance to move the stack down to avoid clobbering that data, but the safest way I could think of to do this was to just set up thread local storage to save registers to.

Reply:
Also FYI the fs/gs relationship is reversed on macOS, similar to Windows. This is just accessing some pre-allocated TLS slots to store data. The equivalent of what you're doing on Linux in your other PR would be storing the guest TLS in FS (which I suppose makes things even simpler, assuming macOS lets you actually modify it).

Reply:
Would be good to check (on windows I believe red-prig found that FS is getting reset by the OS on context switches, making it impractical to use)

Reply:
Sure thing